modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java

```diff
@@ -213,6 +222,71 @@ boolean evaluatePolicies(V value) {
         return true;
     }
 
+    /**
+     * ConcatenatedIterables which combines cache iterables and supports remove() functionality as well if underlying
+     * iterator supports it.
+     * @param <K> Type of key.
+     */
+    class ConcatenatedIterables<K> implements Iterable<K> {
+
+        final Iterable<K>[] iterables;
+
+        ConcatenatedIterables(Iterable<K>[] iterables) {
+            this.iterables = iterables;
+        }
+
+        @SuppressWarnings({ "rawtypes", "unchecked" })
+        @Override
+        public Iterator<K> iterator() {
+            Iterator<K>[] iterators = (Iterator<K>[]) new Iterator[iterables.length];
+            for (int i = 0; i < iterables.length; i++) {
+                iterators[i] = iterables[i].iterator();
+            }
+            return new ConcatenatedIterator<>(iterators);
+        }
+
+        class ConcatenatedIterator<T> implements Iterator<T> {
+            private final Iterator<T>[] iterators;
+            private int currentIteratorIndex;
+            private Iterator<T> currentIterator;
+
+            public ConcatenatedIterator(Iterator<T>[] iterators) {
+                this.iterators = iterators;
+                this.currentIteratorIndex = 0;
+                this.currentIterator = iterators[currentIteratorIndex];
+            }
+
+            @Override
+            public boolean hasNext() {
+                // Check if the current iterator has next element
+                while (currentIterator.hasNext()) {
+                    return true;
+                }
+                // If the current iterator is exhausted, switch to the next iterator
+                currentIteratorIndex++;
+                if (currentIteratorIndex < iterators.length) {
+                    currentIterator = iterators[currentIteratorIndex];
+                    // Check if the switched iterator has next element
+                    return hasNext();
```

Comment:
Rather than making this a recursive call, this whole method could be:

```
while (!currentIterator.hasNext()) {
  currentIteratorIndex++;
  if (currentIteratorIndex == iterators.length()) {
    return false;
  }
  currentIterator = iterators[currentIteratorIndex];
}
return true; // Only get here if `currentIterator.hasNext()` returned `true` in the `while`
```